Source/WebCore/rendering/GridMasonryLayout.cpp

```diff
@@ -0,0 +1,143 @@
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include "config.h"
+#include "GridMasonryLayout.h"
+
+#include "rendering/style/GridPositionsResolver.h"
+
+namespace WebCore {
+
+void GridMasonryLayout::performMasonryPlacement(const HashMap<RenderBox*, GridArea>& firstTrackItems, const Vector<RenderBox*> itemsWithDefiniteGridAxisPosition, const Vector<RenderBox*> itemsWithIndefinitePosition, GridTrackSizingDirection masonryAxisDirection)
+{
+    m_masonryAxisDirection = masonryAxisDirection;
+    m_gridAxisDirection = m_masonryAxisDirection == ForRows ? ForColumns : ForRows;
+    m_masonryAxisGridGap = m_renderGrid.gridGap(m_masonryAxisDirection);
+    m_gridAxisTracksCount = m_renderGrid.numTracks(m_gridAxisDirection);
+    m_runningPositions = Vector<LayoutUnit>(m_gridAxisTracksCount);
+
+    // 2.4 Masonry Layout Algorithm
+    addItemsToFirstTrack(firstTrackItems);
+    placeItemsWithDefiniteGridAxisPosition(itemsWithDefiniteGridAxisPosition);
+    placeItemsWithIndefiniteGridAxisPosition(itemsWithIndefinitePosition);
+}
+
+void GridMasonryLayout::addItemsToFirstTrack(const HashMap<RenderBox*, GridArea>& firstTrackItems)
+{
+    for (auto& itemItr : firstTrackItems) {
+        auto* item = itemItr.key;
+        auto gridArea = itemItr.value;
+
+        insertIntoGridAndLayoutItem(item, gridArea);
+    }
+}
+
+void GridMasonryLayout::placeItemsWithDefiniteGridAxisPosition(const Vector<RenderBox*>& itemsWithDefiniteGridAxisPosition)
+{
+    for (auto* item : itemsWithDefiniteGridAxisPosition) {
+        auto itemSpan = m_renderGrid.currentGrid().gridItemSpan(*item, m_gridAxisDirection);
+
+        ASSERT(!itemSpan.isIndefinite());
+
+        auto initialTrackSpan = GridSpan::translatedDefiniteGridSpan(0, 1);
+        itemSpan.translate(m_renderGrid.currentGrid().explicitGridStart(m_gridAxisDirection));
+
+        GridArea area = m_masonryAxisDirection == ForRows ? GridArea(initialTrackSpan, itemSpan) : GridArea(itemSpan, initialTrackSpan);
+        insertIntoGridAndLayoutItem(item, area);
+    }
+}
+
+void GridMasonryLayout::placeItemsWithIndefiniteGridAxisPosition(const Vector<RenderBox*>& itemsWithIndefinitePosition)
+{
+    for (auto* item : itemsWithIndefinitePosition)
+        insertIntoGridAndLayoutItem(item, nextMasonryPositionForItem(item));
+}
+
+void GridMasonryLayout::insertIntoGridAndLayoutItem(RenderBox* child, const GridArea& area)
+{
+    m_renderGrid.currentGrid().insert(*child, area);
+
+    if (m_gridAxisDirection == ForColumns)
+        child->setOverridingContainingBlockContentLogicalWidth(m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, ForColumns));
+    else
+        child->setOverridingContainingBlockContentLogicalHeight(m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, ForRows));
```

Comment:
... and this could be:

`child->setOverridingContainingBlockContentLogicalWidth(m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, oppositeGridAxisDirection(m_gridAxisDirection)));`

Reply:
Note that ForColumns overrides the logical width, and ForRows overrides the logical height.

Reply:
Oh, right! So why the conditional at all? Wouldn't it just be:

`child->setOverridingContainingBlockContentLogicalWidth(m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, m_gridAxisDirection));`